src/coreclr/jit/copyprop.cpp

```diff
@@ -190,7 +202,129 @@ bool Compiler::optCopyProp(
 
         if (newLclDefVN != lclDefVN)
         {
-            continue;
+            // If the definition is a phi def, and is defined in a block that has a backedge,
+            // we may be able to prove equivalence with some other phi def in that block, despite
+            // the fact that the VNs do not match.
+            //
+            // This happens because VN is one-pass and doesn't know the VNs for backedge PhiArgs,
+            // so the VNs for phi defs for loop entry blocks is always a novel opaque VN. But
+            // we can query those backedge VNs after VN is done, and if all Phi Arg VNs match, then
+            // the two Phi Defs are equivalent and could have had the same VN.
```

Comment:
I'm curious what some of these cases look like... the fact that we have a phi in the header should mean we saw a def inside the loop, so in this case that def has to be redefining the local to have the same value as when we come from outside the loop. I would (naively, maybe?) expect that to be rare.

Reply:
It's defining a local to have the same value as some other local at the start of a loop. So yes both have the same value on initial entry and end up with the same value along the backedge. https://github.com/dotnet/runtime/issues/95645#issuecomment-2204406558 has some notes.

Reply:
Likely the fallout from of pre-increment or similar, the code modifies a value and then uses the pre-modified value... eg

https://github.com/dotnet/runtime/blob/21cde6987f984cec8556b112ec49ca02630a1fa6/src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.Enumerator.cs#L62-L65

This requires a temp in our IR. Morph and loop inversion conspire to make both old and new value live around the loop.